feature(global): auth?!?! USER AUTHENTICATION?!?!?!?!

ollibowers opened the pull request:
todo after merge (martin didnt want me to create more diff):
- **add a logout button on degree-wizard**
- react query key refactoring
- docstrings (will do this after approvals as itll cause annoying changes)
- tests? tests? tests? lol no later
- couple our session storage more closely with the csesoc auth
- remove the last few user axios requests from certain fronteng pages like SettingsMenu
- do a lil user route refactoring so they actually all directly use the database helpers i have created
- other stuff i have noted in minutes
- anything that is `TODO-OLLI(pm)` (pm stands for post auth merge)
- setup docker containers for production
  - this involves also dropping the redis stack to not include gui
- confluence docs baby
- figure out token batch deleting
- can probably do some preliminary error checking on the success login page for expected query params

# Changes (will format nicer in abit)

## backend changes
- removed nodemon from backend docker, in favour of uvicorn watch
- removed init-mongo from the main stack, in favour of just doing this step before the uvicorn start
- added new flags for runserver.py
- refactored the database code to all be in one spot
  - see backend/server/db
  - moved connections into here as well
  - created helper functions ontop the user related queries
  - dont clear users anymore on startup
  - cleaned up mongo setup and handling in general
- refactored the user database
  - now you have invariants
  - also has full mongo validation for all three user related collections
  - users now are either setup or not, no default users anymore
- added a bunch of request and validation helpers for the oidc auth code flow
- added redis and built out a session storage layer ontop of oidc
  - look in auth_utility/sessions
  - includes a set of interface functions that abstract the db helpers
  - supports guest sessions and fedauth (oidc) sessions
  - redis maps session token -> sid & uid
    - and is not persisted
  - mongo collections map;
    - refresh token -> sid
    - sid -> session info
    - uid -> user info
- created some middlewares for getting uid into routes
- converted all user routes to use the uid middleware
- underwrote the user routes to use adapters for now
  - for converting the db models into the TypedDicts
  - will do further conversion and remove the TypedDicts after merge
- summer term is false by default again
- added lifespan manager for future use (could probs remove)
- now requires a sessionsdb env file too
- changed workspace vscode settings so not auto run pytests
- changed run_app.py to use new docker containers
  - also made its environment more explicit. Before `load_dotenv` was just modifying the global environment, meaning the `get_x_env()` calls were a bit redundant
- broke tests (half way through fixing)
  - they atleast run now locally

## frontend changes
- now have token boundaries which handle redirecting based on user state
- added a identity provider which handles refreshing of tokens on page load
  - also auto refreshes just based on interval, other solutions have been explored and will continue to be
- added logout page
- reworked login page
- moved identity and token into its own redux slice
  - SHOULD NOT BE REDUX PERSISTED
- created useToken hook ontop of useSelector to ensure a defined token
  - also has a mode that allows undefined, still better than selector
- converted all request helpers to take in token as param
- removed any google auth related stuff
- converted all queries to use token
- changed where some links go (course-selector is now the first page for setup users)
- added degree wizard as a topbar page since it should be real page now
  - hamburger menu coming soon
- changed up degree wizard to work with new setup states
  - thus removed isComplete field from users
- landing page also now knows where to redirect you too based on isSetup
- error404 page dont show header anymore (since the cart required auth)
- created the required request helpers for auth
- created a withAuthorization request helper for formatting token header
- deleted some dead request helpers
- broke tests



roast me, burn me at the stake

(dm me if u need env files (you will need new env files), ill distribute soon)

ollibowers committed:
wip: setup redirecting and dummy token validation

ollibowers committed:
wip: i believe i can do the full auth flow now

ollibowers committed:
wip: playing around with oidc

ollibowers committed:
wip: still working on authentication, oidc engulfs my mind

this is fun but i have spent hours on this

ollibowers committed:
wip: working on integrating all parts together

ollibowers committed:
wip: started working on session middleware

ollibowers committed:
wip: refactored requests helpers a little bit

ollibowers committed:
wip: got some semblence of sessions working

ollibowers committed:
feat: setup better errors for oidc and session

ollibowers committed:
wip: playing with cookies

ollibowers committed:
wip: ahhhhhhh

ollibowers committed:
wip: started getting somewhere with token pair flow

ollibowers committed:
wip: hooked up oidc checking/refreshing to the identity and userinfo

ollibowers committed:
wip: worked on identity fetching and more route protecting

ollibowers committed:
wip: made route to generate auth url

ollibowers committed:
wip: login url is now generated by backend

ollibowers committed:
wip: now verifying at_hash

ollibowers committed:
wip: setup some dummy setup checking

ollibowers committed:
fix: remove levenstein

ollibowers committed:
wip: fixup after merge

ollibowers committed:
wip: started refactor for new auth flow

ollibowers committed:
wip: redesigned the session interface

ollibowers committed:
wip: brainstorming new mongo and redis structure

ollibowers committed:
wip: changed sids to uuids

ollibowers committed:
wip: refactored session interface to check replay attack earlier

ollibowers committed:
wip: started rewriting auth routes - identity

ollibowers committed:
wip: wrote login route

ollibowers committed:
wip: rewrote rest of the main auth routes

ollibowers committed:
wip: the rewrite base flow works omg

ollibowers committed:
fix: now deletes state cookie if invalid auth code given

ollibowers committed:
wip: better error handling for logout

ollibowers committed:
wip: address most todos in oidc/requests.py

ollibowers committed:
wip: setup redis docker

ollibowers committed:
wip: hooked up the session token part to redis

ollibowers committed:
wip: started to refactor users db

ollibowers committed:
wip: added session related collections

ollibowers committed:
chore: fix up packages after rebase

ollibowers committed:
wip: converted over refresh token storage to mongo

ollibowers committed:
wip: finished converting all to mongo (first step)

ollibowers committed:
fix: didnt realise pagination happened on ft search ugh i hate this

ollibowers committed:
feat: ttl on redis session tokens wooo

ollibowers committed:
feat: indexes on session collections mongo

ollibowers committed:
feat: converted sids back to uuids

ollibowers committed:
feat: fixed up validation for new user storage

ollibowers committed:
wip: started integrating guest sessions

ollibowers committed:
wip: started to refactor storage interface

ollibowers committed:
wip: culled some interface functions

ollibowers committed:
feat: GUEST SESSIONS (not in user layer yet)

ollibowers committed:
wip: fixed up model props in anticipation for user overhall

ollibowers committed:
wip: added user validation schema and collection typing

ollibowers committed:
feat: created new user database helpers

ollibowers committed:
wip: remove uid from models temporarily and fix pylint

ollibowers committed:
wip: added new user storage setup on session creation

ollibowers committed:
wip: db refactor - refactored out the redis connection

ollibowers committed:
wip: db refactor - deleted storage.py and created all da helper files

ollibowers committed:
wip: db refactor - remove cringe db helper prefixes

ollibowers committed:
wip: db refactor - moved out mongo conns into new file

ollibowers committed:
wip: db refactor - no more database.py

ollibowers committed:
wip: db refactor - started combining init-mongo and init-sessionsdb

ollibowers committed:
wip: db refactor - combined init sessionsdb

ollibowers committed:
fix: remove nanoid and bring back old package-lock

ollibowers committed:
wip: user routes refactor - underwrote old set and get

ollibowers committed:
wip: user routes refactor - hooked up guest login and started with fixing edge cases of underwrite

ollibowers committed:
wip: user routes refactor - fixed bugs with degree wizard and db helpers update success checks

ollibowers committed:
wip: user routes refactor - fix for None marks

ollibowers committed:
wip: frontend token refactor - re-setup guest login and id providing and refreshing

ollibowers committed:
wip: frontend token refactor - created new redux slice for identity

ollibowers committed:
wip: frontend token refactor - fixed any straggling old redux usage

ollibowers committed:
wip: frontend token refactor - token param - addToUnplanned & removeCourse

ollibowers committed:
wip: frontend token refactor - token param - setPlanned & setUnplanned & unschedule & unscheduleAll & ignoreFromProg

ollibowers committed:
wip: frontend token refactor - token param - removeAll & validateTermPlanner

ollibowers committed:
wip: frontend token refactor - token param - setupDegree & resetDegree

ollibowers committed:
wip: frontend token refactor - token param - search & updateMark & validateCTF & toggleLocked

ollibowers committed:
wip: frontend token refactor - token param - getUser

ollibowers committed:
wip: frontend token refactor - token param - getUserDegree

ollibowers committed:
wip: frontend token refactor - token param - getUserPlanner & getUserCourses & setIsComplete

ollibowers committed:
wip: frontend token refactor - completely remove getToken

ollibowers committed:
wip: frontend token refactor - setup identity provider and playing around with different refresh methods

ollibowers committed:
wip: frontend token refactor - reworked token state checking and RequireToken

ollibowers committed:
wip: frontend token refactor - Wrapped routes in RequiredToken and PreventToken

ollibowers committed:
wip: frontend token refactor - fixed setup and reset on frontend

ollibowers committed:
wip: frontend token refactor - fixed PreventToken temporarily

ollibowers committed:
wip: frontend token refactor - remove Auth.tsx

ollibowers committed:
wip: frontend token refactor - add back pagetemplate to Login.tsx

ollibowers committed:
wip: frontend token refactor - create Logout page and fix up some query client invalidation

ollibowers committed:
wip: frontend token refactor - fix bug with refreshing again

ollibowers committed:
Merge branch '331-staging' into olli/misc/validate-tokens

ollibowers committed:
wip: fix up models after pydantic upgrade

ollibowers committed:
fix: unplanned courses have null for unplanned

ollibowers committed:
wip: moved all tokens to be passed via headers

ollibowers committed:
feat: create connect function for mongodb conn

ollibowers committed:
feat: replaced nodemon with uvicorn reload for better control of what reloads

ollibowers committed:
feat: effectively removed init-mongo

ollibowers committed:
fix: removed old collections

ollibowers committed:
feat: new straight to uid route dependency

ollibowers committed:
wip: minor route rename

ollibowers committed:
fix: get rid of  field since this is now inferred by the shape of the user

ollibowers committed:
fix: added WWW-Authenticate headers to 401 errors

ollibowers committed:
feat: delete all guest data on logout

ollibowers committed:
feat: added dev flag to runserver.py so production dont reload

ollibowers committed:
chore: relabeled all my todos lol

ollibowers committed:
chore: forgor some

ollibowers committed:
feat: setup secure cookies

ollibowers committed:
fix: search bar quick add buttons not updating

ollibowers committed:
feat: redis password

ollibowers committed:
fix: fixed up login success page (i think idk auth is down)

ollibowers committed:
chore: relabelled todos so i actually know when i need to do what

ollibowers committed:
fix: dict() -> model_dump()

ollibowers committed:
fix: removed the token state route in place for isSetup

ollibowers committed:
wip: changed the isSetup error handling to use throwOnError

ollibowers committed:
wip: added userinfo validation p1

ollibowers committed:
wip: fixed logout redirect and pondered on things

ollibowers committed:
wip: playing around with diff refresh methods

ollibowers committed:
wip: removed whacky thunks for refreshing

ollibowers committed:
wip: change secure cookie prefix and trying out samesite strict

ollibowers committed:
Merge branch '331-staging' into olli/misc/validate-tokens

ollibowers committed:
deps: fix after merge

ollibowers committed:
fix: cookies now get deleted on secure mode

ollibowers committed:
fix: rename session errors

ollibowers committed:
wip: investigated sid and logout again

ollibowers committed:
wip: more todo culling and cleanup

ollibowers committed:
wip: converting over oidc config to be dynamic

ollibowers committed:
fix: summer term off by default

ollibowers committed:
fix: move oidc config into helper (kinda mid sol rn)

ollibowers committed:
fix: changed logout to be post

ollibowers committed:
wip: started working out the redirect behaviour of isSetup 401

ollibowers committed:
fix: useToken now can actually hit error boundary

ollibowers committed:
fix: temporarily solved the redirection 401 conundrum

ollibowers committed:
fix: delete init-mongo from docker-compose

ollibowers committed:
fix: removed some debug token help

ollibowers committed:
feat: made logout a bit more robust i guess

Peedee2002 commented on backend/init-mongo.dockerfile:
```diff
@@ -2,7 +2,7 @@
 # overwriting every time.
 
-FROM python:3.11.5
+FROM python:3.12.3-slim
```
Nit: you can use alpine here to make it a bit less intensive. Sometimes that fucks up arm users, so make sure you test it on an arm machine too.

ollibowers replied:
oooo ya i looked into this a while back, and read it actually kinda sucks for anything with wheels https://www.reddit.com/r/docker/comments/g5hb93/alpine_vs_pythonslim_for_deploying_python_data/ (and doesnt that include levenstein?)


(init-mongo is dead btw so this would only really be for ci-backend or whatever)

Peedee2002 commented on backend/server/db/helpers/models.py:
```diff
@@ -0,0 +1,115 @@
+# https://github.com/pydantic/pydantic/issues/1223#issuecomment-1152323275
+class PartialUserStorage(BaseModel):
+    # for batch user storage update
+    # TODO-OLLI(pm): remove this and use keyword args since they cannot be None
```
There is a way for uvicorn to handle this 👀
We were forced to do this because BaseModel didn't allow "NotRequired"

ollibowers replied:
how do u mean?

Peedee2002 reviewed (commented):
Very cool, will review on my PC later

imagine-hussain commented:
Holy hell

imagine-hussain commented on backend/runserver.py:
```diff
@@ -4,25 +4,45 @@
-    if args.overwrite:
+    if args.overwrite or args.dev:
+        # TODO-OLLI(pm): abstract this, and remove these local imports once we have proper connection handling
+        # TODO-OLLI(pm): also do overwrite checks for dev mode using a timestamp
```
What is the `overwrite check` consisting of?

ollibowers replied:
this is in relation to overwriting course data on every startup, waiting 30 seconds agitates me... so mayb store timestamps on scrape, and do a quick query to check if we actually need to overwrite the archives and course data. (would only be if `dev=True`, prod should overwrite regardless)

imagine-hussain replied:
I see. I would rename to `overwrite_data_on_startup`

ollibowers replied:
oh wait if you mean the args (i thought you meant the comments), then yea `overwrite` was for always overwriting data and has been there for ages, `dev` is used for when your in dev, so has the hot reload and overwrite (or as mentioned above, hopefully smart overwrite)...

but yea, will rename

imagine-hussain commented on backend/runserver.py:
```diff
@@ -4,25 +4,45 @@
+        host='0.0.0.0',
+        lifespan="on",
+        reload=args.dev,
+        reload_excludes="*.json" if args.dev else None,
```
> Set glob patterns to exclude while watching
                                  for files. Includes '.*, .py[cod], .sw.*,
                                  ~*' by default; these defaults can be
                                  overridden with `--reload-include`. This
                                  option has no effect unless watchfiles is
                                  installed.

We lose the the exclusion of dot and swap files in dev, which we probably don't want to do.

ollibowers replied:
oh super interesting

ollibowers replied:
what are these files for?

imagine-hussain replied:
The `~*` files are swapfiles that are sometimes created by editors as a backup of a buffer. Often for things like preserving some state to undo to after the editor exits or, for reverting after failed file writes.

ollibowers replied:
(looks like it combines them, will revisit if we find an issue)

imagine-hussain commented on backend/server/db/helpers/models.py:
```diff
@@ -0,0 +1,115 @@
+# https://github.com/pydantic/pydantic/issues/1223#issuecomment-1152323275
+class PartialUserStorage(BaseModel):
+    # for batch user storage update
+    # TODO-OLLI(pm): remove this and use keyword args since they cannot be None
```
Is this comment
1. The fields should no longer be none?
2. The arguments passed in should not be allowed to be `None` but, the argument can be omitted?

ollibowers replied:
Im still on fence with this.
The aim was to allow just passing in Degree or Courses or a combo of any... But with Pydantic, to indicate a not required field, you must also allow that field to be `None`. Because of this, In Pydantics `model_dump` with `exclude_unset`, something like `PartialUserStorage(courses=...)` would convert to a dict with only a `courses` key, BUT, something like `PartialUserStorage(courses=..., degree=None)` converts to a dict including the `degree` field (BAD). An alternative is using the `exclude_none` param in `model_dump`, BUT an adverse side effect is that this applies over the whole dump, where we want to keep `None` in some fields... So with this Pydantic way, I have to filter out the `None` fields at the top level after deserializing...

So this comment is more just saying "I don't actually want there to be `None` fields". 

With keyword args, you can do something like `def func(*, a: int, b: int): ...` which truly allows you to have optional fields or params that are not allowed to be given as `None`, thus removing any unclear behaviour.

ollibowers replied:
but tbh, i think the post deserialization filter is just cleanest, and gives the type validation i want, whereas keyword args would not (specifically for the courses dict), without using the pydantic validation decorators (which are more added code complexity)

imagine-hussain commented on backend/server/db/helpers/session_tokens.py:
```diff
@@ -0,0 +1,75 @@
+from redis.commands.search.query import Query
+from .models import SessionID, SessionToken, SessionTokenInfoModel
+
+# FT.CREATE idx:uid ON HASH PREFIX 1 "stoken:" NOOFFSETS NOHL NOFIELDS NOFREQS STOPWORDS 0 SCHEMA uid TAG CASESENSITIVE
+# FT.SEARCH idx:uid "@uid:{z5362383}" NOCONTENT VERBATIM
+# FT.SEARCH idx:sid "@sid:{bf362dd3\\-046a\\-49e4\\-8d63\\-fd379f06a40f}" NOCONTENT VERBATIM
+# HMGET stoken:Nj-C2n8JYS4imxlVn7WhSX5Pa0uRZ1awa1w_YA-vnN4jMXXqqKVb7HW9rNhfGako sid uid
```
Are these meant to still be here? What do they do?

ollibowers replied:
oh this is just for self reference later, since its kinda hard to figure out these queries

ollibowers replied:
could probs remove

Peedee2002 replied:
if these are helpful, probably make them real functions so you can delete them from the comments

ollibowers replied:
they already are (the functions in the file are implementations of them), was moreso keeping them around coz i use them in debugging... I will remove after merge when we revisit the indexing (i want to get rid of redis search its not appropriate)

imagine-hussain commented on backend/server/db/helpers/sessions.py:
```diff
@@ -0,0 +1,100 @@
+
+    return res.matched_count == 1
+
+def delete_all_sessions(sid: SessionID) -> bool:
+    res = sessionsNewCOL.delete_many({ "sid": sid })
+    return res.deleted_count > 0
```
Makes more sense for this function to return the count. Callers that were going to do check the return value can do their own checks for `> 0` - or even keep the assert / condition the same.

ollibowers replied:
this is actually interesting, as there should only be one session per sid (as is enforced by a unique sid index)... I think I had the delete_many for sanity.

Should I convert to delete_one and change it to return `res.deleted_count == 1`?

imagine-hussain replied:
If we are sure that there will be only one - which I think we are able to enforce using the database itself right?
In that case, `delete_one` should be fine / preferred

ollibowers replied:
agreed now

imagine-hussain commented on backend/server/db/helpers/users.py:
```diff
@@ -0,0 +1,192 @@
+
+# TODO-OLLI(pm): decide if we want to remove type ignores by constructing dictionaries manually
+
+def get_user(uid: str) -> Optional[Union[NotSetupUserStorage, UserStorage]]:
+    res = usersNewCOL.find_one({ 'uid': uid })
+    if res is None:
+        return res
+
+    return UserStorage.model_validate(res) if res["setup"] else NotSetupUserStorage.model_validate(res)
+
+def get_user_degree(uid: str) -> Optional[UserDegreeStorage]:
+    res = get_user(uid)
+    return res.degree if res is not None and res.setup is True else None
+
+def get_user_courses(uid: str) -> Optional[UserCoursesStorage]:
+    res = get_user(uid)
+    return res.courses if res is not None and res.setup is True else None
+
+def get_user_planner(uid: str) -> Optional[UserPlannerStorage]:
+    res = get_user(uid)
+    return res.planner if res is not None and res.setup is True else None
+
+def user_is_setup(uid: str) -> bool:
+    res = usersNewCOL.find_one({ 'uid': uid, 'setup': { "$eq": True } })
+    return res is not None
```
I think optimal API is

```
val get_user : uid -> Optional[Union[NotSetupUserStorage, UserStorage]]
val get_user_validate : uid -> Optional[UserStorage]
```

Then, for devs that want access to the `degree`, `planner` fields, etc

```
user = get_user_validated(id)
if user is None: raise ...

# The following fields should now be non-null
use user.program
use user.degree
```

ollibowers replied:
good point, thats how it ends up actually working, since everything goes through `get_setup_user` in the routes (which auto throws 403), and the courses/degree/planner is just indexed off of that.

I probably wrote these before i realised that

ollibowers replied:
i fixed this

imagine-hussain reviewed (changes requested):
I've had a look at some op the backend configs. Mostly looks good. Some comments about API; will review in more detail soon.

ollibowers committed:
fix: forgot to remove a debug error in landingpage im dumb as hell

ollibowers committed:
fix: oop forgot some more debug comments in IdentityProvider

ollibowers committed:
fix: made the environment capture more explicit for run_app.py

ollibowers committed:
fix: cleanup auth.py logout route a lil

ollibowers committed:
fix: bug with double initial refresh and csesoc logins

ollibowers committed:
fix: changed logout to use hard navigation for data clearing

ollibowers committed:
fix: removed token playground

ollibowers committed:
fix: deleted some unneeded user db helpers

ollibowers committed:
feat: rewrote mongo setup a bit so that it doesnt drop users by default

ollibowers committed:
feat: moved collection names to constants and renamed the dyn cols

ollibowers committed:
feat: redis now gets cleared on startup

Peedee2002 commented:
@ollibowers your docker compose is failing. This is because you changed the way env files worked.
https://github.com/devsoc-unsw/circles/blob/6c49c0cf609e20e6c380ef96c05012daac56f7aa/.github/workflows/ci.yml#L78
Have a look here to make a script which does create fake creds that work.

ollibowers commented:
> @ollibowers your docker compose is failing. This is because you changed the way env files worked.
> 
> https://github.com/devsoc-unsw/circles/blob/6c49c0cf609e20e6c380ef96c05012daac56f7aa/.github/workflows/ci.yml#L78
> 
> Have a look here to make a script which does create fake creds that work.

sory yea i didn't really understand the ci, i will fix

ollibowers committed:
fix: made some mongo setup funcs exportable for testing purposes

ollibowers committed:
fix: added redis reset to test clear helper

ollibowers committed:
fix: tests can run locally now given you have env files

ollibowers committed:
feat: backend can now run without fedauth credentials

imagine-hussain commented on backend/server/db/helpers/models.py:
```diff
@@ -0,0 +1,115 @@
+    access_token: str           # most recent access token
+    raw_id_token: str           # most recent id token string
+    refresh_token: str          # most recent refresh token
+    validated_id_token: dict    # most recent valid id token object
```
Are we able to get any stronger typing on this, than just `dict`?

ollibowers replied:
I do actually have strong types for this in the oidc packages...

My reasoning at the time for doing this was firstly that the expected fields can change afaik, so I didn't want to put strict validation on it in the mongo validations, thus neither here since these models are meant to be representation of the database contents. I also felt like due to this, it could couple the session abstraction to tightly with our expected oidc config. (like ideally the database handling just knows its a dict, then its the oidc layer is responsible for knowing its shape)

BUT that being said, doing this raw dict typing leads to casts being dropped around the place, so I am not sure I agree with my reasoning anymore. 
AND there are minimum set of fields that should always exist, so maybe I could figure out some base level of typing atleast and let there be extra fields. 
ALSO maybe since this session shape is specifically for the csesoc fedauth, that coupling is actually not a bad thing. Worst case, the fields change, we see some validation errors, change our validation shape, and drop all the old sessions which is completely fine.

not too sure tho, thoughts?

imagine-hussain replied:
If we couple this to the CSE fedauth, then we can get validation failures in testing

ollibowers replied:
good point, i will change

imagine-hussain commented on backend/server/db/helpers/models.py:
```diff
@@ -0,0 +1,115 @@
+    uid: str                      # for validation and back lookup
+    oidc_info: SessionOIDCInfoModel
+    curr_ref_token: RefreshToken  # the most recent refresh token, only one that should be accepted
+    type: Literal['csesoc']       # ensure that this can get parsed correctly
```
quite nice :)

imagine-hussain replied:
I do like storing the type like this and letting that be verified but, I don't like that there is no connection between the typeof `SessionInfoModel::type` and `NotSetupSessionModel::type`

Is there a way to create a relation as follows:
```
type NotSetupMarker = Literal['notsetup']
type CSESOCMarker = Literal['csesoc']
...

class SessionInfoModel(...):
  type: Literal[SessionInfoType[CSESOCMarker]]
``

imagine-hussain commented on backend/server/db/helpers/refresh_tokens.py:
```diff
@@ -0,0 +1,41 @@
+        # token already existed
+        return False
+
+def delete_all_refresh_tokens(sid: SessionID) -> None:
+    refreshTokensNewCOL.delete_many({ "sid": sid })
```
Allow this to return the number of refresh tokens deleted.

ollibowers replied:
real af

imagine-hussain commented on backend/server/db/helpers/session_tokens.py:
```diff
@@ -0,0 +1,75 @@
+
+    return True
+
+def delete_all_tokens(sid: SessionID) -> None:
+    # FT.SEARCH idx:sid "@sid:{bf362dd3\\-046a\\-49e4\\-8d63\\-fd379f06a40f}" NOCONTENT VERBATIM
+    # TODO-OLLI(pm): functionize this
+    # look up all the keys, do in patches of DELETE_BATCH_SIZE
+    while True:
+        all_matches = sdb.ft("idx:sid").search(
+            query=Query(f"@sid:{{{sid.hex}}}").no_content().verbatim().paging(0, DELETE_BATCH_SIZE),
+        )
+
+        # the type bindings are wrong, the result has the shape
+        #  {'attributes': [], 'total_results': 1, 'format': 'STRING', 'results': [{'id': '...', 'values': [...]}], 'warning': []}
+        assert isinstance(all_matches, dict)
+        if all_matches['total_results'] > 0:
+            # delete them all
+            # NOTE: potentially this wont work after 1GB worth of keys...
+            sdb.delete(*(m["id"] for m in all_matches["results"]))
+
+        if all_matches['total_results'] <= DELETE_BATCH_SIZE:
+            return  # no more to delete
```
I don't think I understand the need for paging here.
Is there a way to delete everything in a single query? Is there anything that ends up breaking if we do that?

ollibowers replied:
Not really unfortunately, by default its paged to 10 results hence why i give a paged amount, and the max you can do is 10k (without changing configs), so I figured I'd just deal with it. (see LIMIT of FT.SEARCH)

Can probs up the constant though.

imagine-hussain commented on backend/server/db/helpers/refresh_tokens.py:
```diff
@@ -0,0 +1,41 @@
+
+from .models import RefreshToken, RefreshTokenInfoModel, SessionID
+
+def get_refresh_token_info(token: RefreshToken) -> Optional[RefreshTokenInfoModel]:
+    info = refreshTokensNewCOL.find_one({ 'token': token })
+
+    if info is None:
+        return None
+
+    exp = int(info["expiresAt"].timestamp())
+    if exp <= int(time()):
+        return None
+
+    return RefreshTokenInfoModel(
+        sid=SessionID(info["sid"]),
+        expires_at=exp,
+    )
```
How do expired tokens get cleaned up?

ollibowers replied:
I have put a ttl index on the documents based on the expiresAt field, which will handle it for us which is crazy cool (see the setup functions)

imagine-hussain commented on backend/server/routers/auth_utility/oidc/errors.py:
```diff
@@ -0,0 +1,127 @@
+
+    def __str__(self) -> str:
+        desc = self.error_description or "No description."
+        return f"{super().__str__()}\n  Description: {desc}"
```
Super nitpicky but, I woudl remove the single space after `\n` and replace it with a `\t` or, just remove the ` ` entirely

ollibowers replied:
true

imagine-hussain commented on backend/server/routers/courses.py:
```diff
@@ -25,6 +25,8 @@
 
+require_uid = HTTPBearerToUserID()
+
 # TODO: would prefer to initialise ALL_COURSES here but that fails on CI for some reason
 ALL_COURSES: Optional[Dict[str, str]] = None
```
Since we are finally getting a proper startup with redis, we should look into getting this fixed

ollibowers replied:
real

ollibowers replied:
(will do in future)

imagine-hussain reviewed (commented):
Reviewed much more of the backend - still some distance to go.

The wrappers around the db / redis code really make me wish that we were able to enforce something like C++'s `[[nodiscard]]` or rust's `#[must_use]`.

However, this doesn't look like something that `mypy` currently supports :c
We can just be extra-careful when it comes to review I suppose.

https://github.com/python/mypy/issues/17069

imagine-hussain commented on backend/server/routers/auth.py:
```diff
@@ -1,14 +1,276 @@
+from .auth_utility.oidc.requests import DecodedIDToken, exchange_and_validate, generate_oidc_auth_url, get_userinfo_and_validate, refresh_and_validate, revoke_token, validate_authorization_response
+from .auth_utility.oidc.errors import OIDCInvalidGrant, OIDCInvalidToken, OIDCTokenError, OIDCValidationError
+
+STATE_TTL = 10 * 60
```
We should be more specific about what this it a TTL for

ollibowers replied:
really true, i will rename it to `AUTH_STATE_COOKIE_TTL` to match the other constants,

imagine-hussain commented on backend/server/routers/auth.py:
```diff
@@ -1,14 +1,276 @@
+                headers={ "WWW-Authenticate": "Bearer", "set-cookie": res.headers["set-cookie"] },
+            ) from e
+        except OIDCTokenError as e:
+            # cant imagine what error this could be, but if it happens, its bad
+            print(e)
```
Do we currently have a way to track any internal server errors that take place in prod?

ollibowers replied:
i dont think so, a proper logger instance and usage is definitely needed (and something I have on the list of post merge todos)

imagine-hussain commented on backend/server/routers/auth_utility/ideas.md:
Just double-checking, this is committed intentionaly?

ollibowers replied:
true i will remove

imagine-hussain commented on backend/server/db/helpers/users.py:
```diff
@@ -0,0 +1,160 @@
+        # most semantically correct
+        return user_is_setup(uid)
+
+    if "courses" in payload and "degree" in payload and "planner" in payload:
+        # enough to declare user as setup
+        payload["setup"] = True
+
+    res = usersNewCOL.update_one(
+        { "uid": uid, "setup": True } if "setup" not in payload else { "uid": uid },
```
Is that ternary the right way around?
In the first if-statement, having the key means that the user is fully set up.
However, in the update, we insert the `setup: True` if we don't have the setup key in the payload

ollibowers replied:
It's kinda confusing, but this is the filter arg for the update (what is used to find the documents we want to update).

You are right, we add the "setup" key into the payload above if they have given us all the other keys required to make a user setup. 
SO, if they haven't given us all the properties we need (and the key is not present), we are only patching certain fields, thus we require our search to look for a document that already setup (or we will be missing fields). 
OTHERWISE, if they have given us all the keys we need, then we can apply the update to the user regardless if they are setup or not. 

So we are doing:
> If we do not have all keys, look for a match that is already setup, otherwise just look for a user

Maybe I can flip the ternary to make this a bit clearer?

imagine-hussain reviewed (changes requested):
Finished review of the backend and added comments.

Peedee2002 commented on .vscode/settings.json:
```diff
@@ -4,7 +4,6 @@
     ],
     "VitestRunner.executionArg": "npx --prefix frontend",
     "python.testing.unittestEnabled": false,
-    "python.testing.pytestEnabled": true,
```
🤔 i thought this worked correctly and made sure you can run your pytest using vscode

ollibowers replied:
it wassss, but it was also causing variables at top level to get evaluated in the test files, (and there are two variables which required route calls, so it was making random route calls everytime i saved which was bugging me)

Peedee2002 commented on backend/dev.dockerfile:
```diff
@@ -25,5 +15,11 @@ COPY . .
+
 # Run the server
-ENTRYPOINT nodemon --exec python3 -u runserver.py
+# https://forums.docker.com/t/docker-run-cannot-be-killed-with-ctrl-c/13108/2
```
HOLY SHIT that is why? I always thought it was random if it cancelled or not. Wild.

ollibowers replied:
bahahahhahahhaha, it is reallllyyy whack, before finding that everywhere i ever read also just said array and string were equivalent

Peedee2002 commented:
There should be some docs https://github.com/devsoc-unsw/circles/blob/dev/backend/README.md that require update

ollibowers commented:
> There should be some docs https://github.com/devsoc-unsw/circles/blob/dev/backend/README.md that require update

good point, i will get this done!

Peedee2002 committed:
update CI for testing with new stuff

Peedee2002 committed:
update documentation

Peedee2002 commented on backend/requirements.txt:
```diff
@@ -1,18 +1,55 @@
+rapidfuzz==3.9.0
+redis==5.0.3
 requests==2.31.0
+six==1.16.0
```
this is a valid decision if you are **really finnicky** about your transitive dependencies. It will just make a lot of noise because you will have to manage these dependencies yourself and will get dependabot on your ass.

ollibowers replied:
yeaaaa ://// 

I was thinking of investigating something like poetry, or some more advanced (but simple) package manager that has a lock file.

ollibowers replied:
i like it

Peedee2002 commented on backend/server/config.py:
```diff
@@ -2,10 +2,10 @@
 ARCHIVED_DATA_PATH = "./data/final_data/archive/processed/"
 
-DUMMY_TOKEN = "token"
+DUMMY_TOKEN = "token"  # TODO: get rid of this
```
yea, this was a mistake 💀 . There is a similar issue in the frontend with "bad_planner" or something. will be good to refactor that at some point.

ollibowers replied:
yaaa dummy data like bad_planner is used a lot in frontend because we never made a standard way of dealing with loading and error query states. 
in my react query followon PR, i will likely unify it all.

for dummy_token, its only used now in testing, so when we fix tests, we can just use guest login route instead.

Peedee2002 committed:
fix readme

Peedee2002 commented on backend/server/routers/user.py:
```diff
@@ -1,59 +1,115 @@
+
+
 from pydantic import BaseModel
+# TODO-OLLI(pm): i think we can get rid of this now since we do not use ObjectId anywhere
```
yea, sounds legit

ollibowers replied:
oki 🫡

Peedee2002 commented on docker-compose.yaml:
```diff
@@ -73,19 +93,8 @@ services:
-    networks:
-      - circles-net
+
+  # TODO-OLLI(pm): idek if this works or how it should work post init-mongo death
```
I doubt it

ollibowers replied:
Oh i actually was thinking how to replace this, since i think ci mainly uses it to get the backend containers to run and cleaned...

I think if we can assume that the containers start empty, then we can just shift the ci to start backend directly (as itll run in dev, which at the moment only drops users). 
But, if we cannot, then i can make a simple new dockerfile that is just the same as the production backend docker (overwrites courses), but uses a new clarg to drop users too...

thoughts?

Peedee2002 replied:
I believe the purpose was something to do with handling the production instances of databases. Not 100% sure tho.

ollibowers replied:
interesting

Peedee2002 reviewed (commented):
a few small comments, damn its been a min since i have seen circles code

ollibowers committed:
fix: cleaned up various return values and files after PR review feedback

- fixed return values of delete session and refresh token helpers
- removed ideas.md file
- changed oidc error string method to use tabs instead of spaces
- renamed state cookie and its ttl constant

ollibowers committed:
fix: droppped 'New' naming from collection variables

Peedee2002 committed:
bump random thingo to remove high vuln

ollibowers committed:
fix: explored better marker typing for models, found out pydantic is broken

zax-xyz commented on frontend/src/components/Auth/RequireToken.tsx:
```diff
@@ -0,0 +1,76 @@
+};
+
+const RequireToken = ({ needSetup }: Props) => {
+  const { token } = useAppSelector(selectIdentity) ?? {};
+
+  const {
+    isPending,
+    data: isSetup,
+    error
+  } = useQuery({
+    queryKey: ['degree', 'isSetup'], // TODO-OLLI(pm): fix this key, including userId
+    queryFn: () => getUserIsSetup(token!),
+    enabled: token !== undefined,
```
thoughts on making a generic `useQueryWithToken` hook or similar that grabs the token and injects it into the `enabled` etc to reduce boilerplate?

ollibowers replied:
this is definitely something we will add after merge!

zax-xyz commented on frontend/src/components/PlannerButton/PlannerButton.tsx:
```diff
@@ -12,7 +13,11 @@ interface PlannerButtonProps {
-  const handleMutation = isAddedInPlanner ? removeCourse : addToUnplanned;
+  const token = useToken();
+
+  const handleMutation = isAddedInPlanner
+    ? (code: string) => removeCourse(token, code)
+    : (code: string) => addToUnplanned(token, code);
```
thoughts on
```suggestion
  const handleMutation = (code: string) => (isAddedInPlanner ? removeCourse : addToUnplanned)(token, code);
```

zax-xyz replied:
or perhaps
```suggestion
  const handleMutation = (code: string) => isAddedInPlanner ? removeCourse(token, code) : addToUnplanned(token, code);
```

ollibowers replied:
Really good point. 
After looking more, I have noticed this is inconsistent in many places, so i might address in the query followup PR.
(These funky mutations were not added in this branch)

zax-xyz commented on frontend/src/components/QuickAddCartButton/QuickAddCartButton.tsx:
```diff
@@ -14,7 +15,11 @@ type Props = {
-  const handleMutation = planned ? removeCourse : addToUnplanned;
+  const token = useToken();
+
+  const handleMutation = planned
+    ? (code: string) => removeCourse(token, code)
+    : (code: string) => addToUnplanned(token, code);
```
same here

ollibowers replied:
ditto

zax-xyz commented on frontend/src/components/ResetModal/ResetModal.test.tsx:
```diff
@@ -20,23 +19,25 @@ describe('ResetModal', () => {
-    ]);
-  });
+  // it('should show modal when degree wizard is complete', async () => {
+  //   await renderWithProviders(<ResetModal open />);
+  //   // TODO: idk what to do with token here
+  //   await setIsComplete('', true);
+  //   expect(screen.getByText('Reset Planner?')).toBeInTheDocument();
+  // });
+
+  // it('should reset state when reset button is clicked', async () => {
+  //   const dummyDispatch = vi.fn();
+  //   useDispatchMock.mockReturnValue(dummyDispatch);
+
+  //   await renderWithProviders(<ResetModal open />);
+  //   // TODO: idk what to do with token here
+  //   await setIsComplete('', true);
+  //   await userEvent.click(screen.getByText('Reset'));
+  //   expect(dummyDispatch.mock.calls).toEqual([
+  //     [{ payload: undefined, type: 'courseTabs/resetTabs' }]
+  //   ]);
+  // });
```
💀

ollibowers replied:
💀

zax-xyz commented on frontend/src/config/store.ts:
```diff
@@ -2,12 +2,14 @@ import { configureStore } from '@reduxjs/toolkit';
   courseTabs: courseTabsReducer,
-  settings: settingsReducer
+  settings: settingsReducer,
+  identity: identityReducer
```
surely add trailing commas to prettier config

ollibowers replied:
real

zax-xyz commented on frontend/src/hooks/useToken.ts:
```diff
@@ -0,0 +1,52 @@
+  const token = useAppSelector(selectToken);
+
+  if (token === undefined && options?.allowUnset !== true) {
+    // This shouldn't occur of useToken is used inside a RequireToken
```
:pwned:
```suggestion
    // This shouldn't occur if useToken is used inside a RequireToken
```

zax-xyz commented on frontend/src/hooks/useToken.ts:
```diff
@@ -0,0 +1,52 @@
+
+  if (token === undefined && options?.allowUnset !== true) {
+    // This shouldn't occur of useToken is used inside a RequireToken
+    throw TypeError('useToken: allowUnset was false when token was undefined.', {
+      cause: { token, allowUnset: !!options?.allowUnset }
+    });
```
thoughts on just putting the cause as part of the error message string?

ollibowers replied:
yes, i will do this!!!!!!

zax-xyz commented on frontend/src/pages/CourseSelector/CourseMenu/CourseMenu.tsx:
```diff
@@ -76,10 +79,8 @@ const CourseMenu = ({ planner, courses, degree }: CourseMenuProps) => {
-      const handleMutation = inPlanner(courseId) ? removeCourse : addToUnplanned;
-      await handleMutation(courseId);
-    },
+    mutationFn: async (courseId: string) =>
+      inPlanner(courseId) ? removeCourse(token, courseId) : addToUnplanned(token, courseId),
```
does this still need to be async?

ollibowers replied:
ditto

zax-xyz commented on frontend/src/pages/DegreeWizard/DegreeWizard.tsx:
```diff
@@ -31,12 +33,30 @@ const DegreeWizard = () => {
+  const resetDegree = useMutation({
+    mutationFn: () => resetUserDegree(token),
+    onSuccess: () => {
+      queryClient.invalidateQueries({
+        queryKey: ['degree']
+      });
+      queryClient.invalidateQueries({
+        queryKey: ['courses']
+      });
+      queryClient.invalidateQueries({
+        queryKey: ['planner']
+      });
```
thoughts on 
```suggestion
      ['degree', 'courses', 'planner'].forEach(key =>
        queryClient.invalidateQueries({ queryKey: [key] });
      );
```

ollibowers replied:
oooo this is also good idea! will do in follow up!

zax-xyz commented on frontend/src/pages/LoginSuccess/LoginSuccess.tsx:
```diff
@@ -0,0 +1,44 @@
+        const identity = await CSELogin(Object.fromEntries(query.entries()));
+
+        dispatch(updateIdentityWithAPIRes(identity));
+        // TODO-OLLI(pm): could use a conditional useQuery too
```
so true

ollibowers replied:
i actually disagree now 😡

zax-xyz commented on frontend/src/pages/TermPlanner/ImportPlannerMenu/ImportPlannerMenu.tsx:
```diff
@@ -1,31 +1,33 @@
     try {
-      await axios.post('planner/unPlannedToTerm', data, { params: { token } });
+      await axios.post('planner/unPlannedToTerm', data, {
+        headers: { ...withAuthorization(token) }
```
thoughts on 
```suggestion
        headers: withAuthorization(token)
```

ollibowers replied:
good pick up, have fixed!

zax-xyz commented on frontend/src/reducers/identitySlice.ts:
```diff
@@ -0,0 +1,45 @@
+    selectIdentity: (state) => state
+  },
+  reducers: {
+    updateIdentity: (state, action: PayloadAction<CompleteIdentity>) => {
+      return {
+        expiresAt: action.payload.expiresAt,
+        token: action.payload.token,
+        userId: action.payload.userId
+      };
+    },
+    updateIdentityWithAPIRes: (state, action: PayloadAction<IdentityResponse>) => {
+      return {
+        expiresAt: action.payload.exp,
+        token: action.payload.session_token,
+        userId: action.payload.uid
+      };
+    },
```
thoughts on 
```suggestion
    updateIdentity: (state, action: PayloadAction<CompleteIdentity>) => ({
      expiresAt: action.payload.expiresAt,
      token: action.payload.token,
      userId: action.payload.userId
    )},
    updateIdentityWithAPIRes: (state, action: PayloadAction<IdentityResponse>) => ({
      expiresAt: action.payload.exp,
      token: action.payload.session_token,
      userId: action.payload.uid
    )},
```

ollibowers replied:
love it

zax-xyz reviewed (changes requested):
lgtm

ollibowers committed:
fix: major issue of spelling raised by zax-xyz

Co-authored-by: Michael Vo <zax@zaxu.xyz>

lhjt commented on .github/workflows/ci.yml:
Very crazy have you been able to test if this is functional?

ollibowers replied:
not yet, have not fully fixed ci yet :c

ollibowers replied:
@lhjt as far as I can tell, the docker container `ci-backend` is only used in our `ci.yml`... 
I hear that it might also be used elsewhere behind the scenes, is this true? Am I good to have editted it?

ollibowers replied:
(i have also now fixed the ci to be in its final state for this PR, and the job now runs, so if you need to check anything 🥺)

ollibowers committed:
fix: cleaned up identity slice reducers and thrown errors on frontend

As per Michael feedback!

ollibowers committed:
fix: removed all the spreads on the withAuthorization, as per Michael feedback :)

ollibowers committed:
Merge branch '331-staging' into olli/misc/validate-tokens

ollibowers committed:
fix: removed suppressed and mostRecentPastTerm to align with #1158

ollibowers committed:
feat: added new cli arg for deleting user data aswell, updated ci backend container to mirror prod but use this

Also finally deleted init-mongo.dockerfile and init-database.py

ollibowers committed:
fix: updated ci to use env script

martanman reviewed (approved):
lgtm

ollibowers committed:
Merge branch '331-staging' into olli/misc/validate-tokens

ollibowers committed:
fix: mypy issues with unsupported typing

zax-xyz reviewed (approved):
trash
